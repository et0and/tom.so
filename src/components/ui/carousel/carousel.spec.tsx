import React from "react";
import { render, screen, fireEvent } from "@testing-library/react";
import "@testing-library/jest-dom";
import {
  Carousel,
  CarouselContent,
  CarouselItem,
  CarouselNext,
  CarouselPrevious,
} from "./carousel";
import { expect, describe, it, vi } from "vitest";

// Mock the embla-carousel-react module using Vitest
vi.mock("embla-carousel-react", () => {
  return {
    default: vi.fn(() => {
      // Return a mock implementation of useEmblaCarousel hook
      // First element is a ref function, second is the API
      return [
        vi.fn(), // carouselRef
        {
          canScrollPrev: vi.fn().mockReturnValue(true),
          canScrollNext: vi.fn().mockReturnValue(true),
          scrollPrev: vi.fn(),
          scrollNext: vi.fn(),
          on: vi.fn(),
          off: vi.fn(),
        },
      ];
    }),
  };
});

describe("<Carousel />", () => {
  it("renders correctly and matches snapshot", () => {
    const { container } = render(
      <Carousel>
        <CarouselContent>
          <CarouselItem>Slide 1</CarouselItem>
          <CarouselItem>Slide 2</CarouselItem>
          <CarouselItem>Slide 3</CarouselItem>
        </CarouselContent>
        <CarouselPrevious />
        <CarouselNext />
      </Carousel>,
    );

    expect(container).toMatchSnapshot();
  });

  it("renders with correct accessibility attributes", () => {
    render(
      <Carousel>
        <CarouselContent>
          <CarouselItem>Slide 1</CarouselItem>
          <CarouselItem>Slide 2</CarouselItem>
        </CarouselContent>
        <CarouselPrevious />
        <CarouselNext />
      </Carousel>,
    );

    // Check carousel has proper aria-roledescription
    const carouselElement = screen.getByText("Slide 1").closest("section");
    expect(carouselElement).toHaveAttribute("aria-roledescription", "carousel");

    // Check slides have proper roledescription
    const slides = screen.getAllByText(/Slide \d/);
    expect(slides).toHaveLength(2);

    // Check navigation buttons have proper text
    expect(screen.getByText("Previous slide")).toBeInTheDocument();
    expect(screen.getByText("Next slide")).toBeInTheDocument();
  });

  it("handles keyboard navigation", () => {
    const { container } = render(
      <Carousel>
        <CarouselContent>
          <CarouselItem>Slide 1</CarouselItem>
          <CarouselItem>Slide 2</CarouselItem>
        </CarouselContent>
        <CarouselPrevious />
        <CarouselNext />
      </Carousel>,
    );

    // Get the carousel section element by attribute
    const carouselElement = screen.getByText("Slide 1").closest("section");
    expect(carouselElement).not.toBeNull();

    // Simulate arrow key presses
    if (carouselElement) {
      fireEvent.keyDown(carouselElement, { key: "ArrowLeft" });
      fireEvent.keyDown(carouselElement, { key: "ArrowRight" });
    }

    // The actual scrolling behavior is handled by the mocked API,
    // so we're just testing that the component doesn't crash when keys are pressed
    expect(container).toMatchSnapshot();
  });

  it("renders with vertical orientation", () => {
    const { container } = render(
      <Carousel orientation="vertical">
        <CarouselContent>
          <CarouselItem>Slide 1</CarouselItem>
          <CarouselItem>Slide 2</CarouselItem>
        </CarouselContent>
        <CarouselPrevious />
        <CarouselNext />
      </Carousel>,
    );

    expect(container).toMatchSnapshot();
  });
});
