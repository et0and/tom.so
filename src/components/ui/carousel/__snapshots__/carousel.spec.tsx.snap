// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Carousel /> handles keyboard navigation 1`] = `
<div>
  <section
    aria-roledescription="carousel"
    class="relative"
  >
    <div
      class="overflow-hidden"
    >
      <div
        class="flex h-full items-center"
      >
        <div
          aria-roledescription="slide"
          class="min-w-0 shrink-0 grow-0 basis-full px-4 flex items-center justify-center"
        >
          Slide 1
        </div>
        <div
          aria-roledescription="slide"
          class="min-w-0 shrink-0 grow-0 basis-full px-4 flex items-center justify-center"
        >
          Slide 2
        </div>
      </div>
    </div>
    <button
      class="items-center justify-center whitespace-nowrap text-sm font-medium ring-offset-white transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-gray-950 focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 dark:ring-offset-gray-950 dark:focus-visible:ring-gray-300 border border-gray-200 bg-white hover:bg-gray-100 hover:text-gray-900 dark:border-gray-800 dark:bg-gray-950 dark:hover:bg-gray-800 dark:hover:text-gray-50 absolute h-8 w-8 rounded-full -left-12 top-1/2 -translate-y-1/2 hidden sm:flex"
    >
      <svg
        class="lucide lucide-arrow-left h-4 w-4"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="m12 19-7-7 7-7"
        />
        <path
          d="M19 12H5"
        />
      </svg>
      <span
        class="sr-only"
      >
        Previous slide
      </span>
    </button>
    <button
      class="items-center justify-center whitespace-nowrap text-sm font-medium ring-offset-white transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-gray-950 focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 dark:ring-offset-gray-950 dark:focus-visible:ring-gray-300 border border-gray-200 bg-white hover:bg-gray-100 hover:text-gray-900 dark:border-gray-800 dark:bg-gray-950 dark:hover:bg-gray-800 dark:hover:text-gray-50 absolute h-8 w-8 rounded-full -right-12 top-1/2 -translate-y-1/2 hidden sm:flex"
    >
      <svg
        class="lucide lucide-arrow-right h-4 w-4"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M5 12h14"
        />
        <path
          d="m12 5 7 7-7 7"
        />
      </svg>
      <span
        class="sr-only"
      >
        Next slide
      </span>
    </button>
  </section>
</div>
`;

exports[`<Carousel /> renders correctly and matches snapshot 1`] = `
<div>
  <section
    aria-roledescription="carousel"
    class="relative"
  >
    <div
      class="overflow-hidden"
    >
      <div
        class="flex h-full items-center"
      >
        <div
          aria-roledescription="slide"
          class="min-w-0 shrink-0 grow-0 basis-full px-4 flex items-center justify-center"
        >
          Slide 1
        </div>
        <div
          aria-roledescription="slide"
          class="min-w-0 shrink-0 grow-0 basis-full px-4 flex items-center justify-center"
        >
          Slide 2
        </div>
        <div
          aria-roledescription="slide"
          class="min-w-0 shrink-0 grow-0 basis-full px-4 flex items-center justify-center"
        >
          Slide 3
        </div>
      </div>
    </div>
    <button
      class="items-center justify-center whitespace-nowrap text-sm font-medium ring-offset-white transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-gray-950 focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 dark:ring-offset-gray-950 dark:focus-visible:ring-gray-300 border border-gray-200 bg-white hover:bg-gray-100 hover:text-gray-900 dark:border-gray-800 dark:bg-gray-950 dark:hover:bg-gray-800 dark:hover:text-gray-50 absolute h-8 w-8 rounded-full -left-12 top-1/2 -translate-y-1/2 hidden sm:flex"
    >
      <svg
        class="lucide lucide-arrow-left h-4 w-4"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="m12 19-7-7 7-7"
        />
        <path
          d="M19 12H5"
        />
      </svg>
      <span
        class="sr-only"
      >
        Previous slide
      </span>
    </button>
    <button
      class="items-center justify-center whitespace-nowrap text-sm font-medium ring-offset-white transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-gray-950 focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 dark:ring-offset-gray-950 dark:focus-visible:ring-gray-300 border border-gray-200 bg-white hover:bg-gray-100 hover:text-gray-900 dark:border-gray-800 dark:bg-gray-950 dark:hover:bg-gray-800 dark:hover:text-gray-50 absolute h-8 w-8 rounded-full -right-12 top-1/2 -translate-y-1/2 hidden sm:flex"
    >
      <svg
        class="lucide lucide-arrow-right h-4 w-4"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M5 12h14"
        />
        <path
          d="m12 5 7 7-7 7"
        />
      </svg>
      <span
        class="sr-only"
      >
        Next slide
      </span>
    </button>
  </section>
</div>
`;

exports[`<Carousel /> renders with vertical orientation 1`] = `
<div>
  <section
    aria-roledescription="carousel"
    class="relative"
  >
    <div
      class="overflow-hidden"
    >
      <div
        class="flex flex-col"
      >
        <div
          aria-roledescription="slide"
          class="min-w-0 shrink-0 grow-0 basis-full py-4 flex items-center justify-center"
        >
          Slide 1
        </div>
        <div
          aria-roledescription="slide"
          class="min-w-0 shrink-0 grow-0 basis-full py-4 flex items-center justify-center"
        >
          Slide 2
        </div>
      </div>
    </div>
    <button
      class="items-center justify-center whitespace-nowrap text-sm font-medium ring-offset-white transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-gray-950 focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 dark:ring-offset-gray-950 dark:focus-visible:ring-gray-300 border border-gray-200 bg-white hover:bg-gray-100 hover:text-gray-900 dark:border-gray-800 dark:bg-gray-950 dark:hover:bg-gray-800 dark:hover:text-gray-50 absolute h-8 w-8 rounded-full -top-12 left-1/2 -translate-x-1/2 rotate-90 hidden sm:flex"
    >
      <svg
        class="lucide lucide-arrow-left h-4 w-4"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="m12 19-7-7 7-7"
        />
        <path
          d="M19 12H5"
        />
      </svg>
      <span
        class="sr-only"
      >
        Previous slide
      </span>
    </button>
    <button
      class="items-center justify-center whitespace-nowrap text-sm font-medium ring-offset-white transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-gray-950 focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 dark:ring-offset-gray-950 dark:focus-visible:ring-gray-300 border border-gray-200 bg-white hover:bg-gray-100 hover:text-gray-900 dark:border-gray-800 dark:bg-gray-950 dark:hover:bg-gray-800 dark:hover:text-gray-50 absolute h-8 w-8 rounded-full -bottom-12 left-1/2 -translate-x-1/2 rotate-90 hidden sm:flex"
    >
      <svg
        class="lucide lucide-arrow-right h-4 w-4"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M5 12h14"
        />
        <path
          d="m12 5 7 7-7 7"
        />
      </svg>
      <span
        class="sr-only"
      >
        Next slide
      </span>
    </button>
  </section>
</div>
`;
